File: backend/app/services/change_detector.py
"""
Change Detection Service
"""
from typing import Dict, List, Optional, Any
from supabase import Client
from app.services.github_service import GitHubService
from app.utils.change_comparison import (
    compare_code_snapshots,
    compare_detection_results,
    should_regenerate_diagram
)
from app.utils.tool_detection import detect_tools
from datetime import datetime

async def detect_repository_changes(
    supabase: Client,
    user_id: str,
    repo_url: Optional[str] = None,
    branch: Optional[str] = None,
    commit_range: Optional[str] = None,
    submission_id: Optional[str] = None,
    diagram_id: Optional[str] = None
) -> Dict[str, Any]:
    """
    Detect changes in a repository since last analysis.
    """
    github_service = GitHubService(supabase, user_id)
    
    # Get old snapshot from database
    old_snapshot = None
    old_detection_result = None
    
    if submission_id:
        # Load submission
        try:
            submission = supabase.table('submissions').select('*').eq('id', submission_id).single().execute()
            if submission.data:
                old_snapshot = submission.data.get('code_snapshot')
                repo_url = submission.data.get('source_meta', {}).get('repoUrl') or repo_url
                branch = submission.data.get('source_meta', {}).get('branch') or branch or 'main'
        except:
            pass
    elif diagram_id:
        # Load diagram
        try:
            diagram = supabase.table('architecture_diagrams').select('*').eq('id', diagram_id).single().execute()
            if diagram.data:
                old_snapshot = diagram.data.get('code_snapshot')
                old_detection_result = diagram.data.get('detection_result')
                repo_url = diagram.data.get('repo_url') or repo_url
                branch = diagram.data.get('branch') or branch or 'main'
        except:
            pass
    
    if not repo_url:
        raise ValueError("repo_url is required")
    
    branch = branch or 'main'
    
    # Get current state
    commit_sha = await github_service.get_commit_sha(repo_url, branch)
    if not commit_sha:
        raise ValueError(f"Could not get commit SHA for {repo_url}/{branch}")
    
    # Get current file SHAs
    if old_snapshot and old_snapshot.get('fileShas'):
        file_paths = list(old_snapshot['fileShas'].keys())
    else:
        # Fetch all files to get current state
        files = await github_service.fetch_repo_files(repo_url, branch)
        file_paths = [f['path'] for f in files]
    
    current_file_shas = await github_service.get_file_shas(repo_url, branch, file_paths)
    
    new_snapshot = {
        'commitSha': commit_sha,
        'fileShas': current_file_shas,
        'createdAt': datetime.utcnow().isoformat()
    }
    
    # Compare snapshots
    code_comparison = compare_code_snapshots(old_snapshot, new_snapshot)
    
    # Compare detection results if applicable
    detection_comparison = {'hasChanges': False}
    architecture_changes = None
    
    if diagram_id and old_detection_result:
        # Re-run detection on changed files
        changed_files = []
        for file_change in code_comparison['filesChanged']:
            # Fetch changed file content
            files = await github_service.fetch_repo_files(repo_url, branch)
            changed_file = next((f for f in files if f['path'] == file_change['path']), None)
            if changed_file:
                changed_files.append(changed_file)
        
        if changed_files or code_comparison['filesAdded']:
            new_detection_result = detect_tools(changed_files)
            detection_comparison = compare_detection_results(old_detection_result, new_detection_result)
            
            architecture_changes = {
                'tools_added': detection_comparison['toolsAdded'],
                'tools_removed': detection_comparison['toolsRemoved'],
                'connections_added': detection_comparison['connectionsAdded'],
                'connections_removed': detection_comparison['connectionsRemoved']
            }
    
    # Build summary
    summary_parts = []
    if code_comparison['commitChanged']:
        summary_parts.append("Commit changed")
    if code_comparison['filesChanged']:
        summary_parts.append(f"{len(code_comparison['filesChanged'])} file(s) modified")
    if code_comparison['filesAdded']:
        summary_parts.append(f"{len(code_comparison['filesAdded'])} file(s) added")
    if code_comparison['filesRemoved']:
        summary_parts.append(f"{len(code_comparison['filesRemoved'])} file(s) removed")
    
    summary = '. '.join(summary_parts) if summary_parts else 'No changes detected'
    
    # Update database
    if submission_id:
        try:
            supabase.table('submissions').update({
                'is_outdated': code_comparison['hasChanges'],
                'last_checked_at': datetime.utcnow().isoformat()
            }).eq('id', submission_id).execute()
        except:
            pass
    elif diagram_id:
        try:
            supabase.table('architecture_diagrams').update({
                'last_checked_at': datetime.utcnow().isoformat()
            }).eq('id', diagram_id).execute()
        except:
            pass
    
    return {
        'has_changes': code_comparison['hasChanges'],
        'commit_changed': code_comparison['commitChanged'],
        'files_changed': [
            {
                'path': fc['path'],
                'old_hash': fc['oldHash'],
                'new_hash': fc['newHash'],
                'status': 'modified'
            }
            for fc in code_comparison['filesChanged']
        ],
        'files_added': code_comparison['filesAdded'],
        'files_removed': code_comparison['filesRemoved'],
        'architecture_changes': architecture_changes,
        'summary': summary,
        'current_commit_sha': commit_sha,
        'old_commit_sha': old_snapshot.get('commitSha') if old_snapshot else None
    }

File: frontend/src/lib/server/services/changeDetector.ts
import type { SupabaseClient } from '@supabase/supabase-js';
import { getUserOctokit } from '../github/getUserOctokit';
import { parseRepoUrl } from '../github/github';
import { analyzeRepository, AnalyzeRepositoryResult } from './analyzeRepository';
import {
	compareCodeSnapshots,
	compareDetectionResults,
} from '../architecture/detectChanges';
import { detectTools } from '../architecture/detectTools';
import type { DetectionResult } from '../architecture/detectTools';
import { getCachedBranch, getCachedCompareCommits } from '../github/cachedOctokit';
import { fetchFilesSmart } from '../github/batchFetch';
import { getRateLimitStatus } from '../github/rateLimiter';

type DetectChangesParams = {
	supabase: SupabaseClient;
	userId: string;
	repoUrl?: string | null;
	branch?: string | null;
	commitRange?: string | null;
	submissionId?: string | null;
	diagramId?: string | null;
};

type FileChange = {
	path: string;
	old_path?: string;
	old_hash?: string | null;
	new_hash?: string | null;
	status: 'modified' | 'renamed' | 'added' | 'removed';
};

export async function detectRepositoryChanges({
	supabase,
	userId,
	repoUrl,
	branch,
	submissionId,
	diagramId,
}: DetectChangesParams): Promise<{
	has_changes: boolean;
	commit_changed: boolean;
	files_changed: FileChange[];
	files_added: string[];
	files_removed: string[];
	files_renamed: Array<{ old_path: string; new_path: string }>;
	architecture_changes?: Record<string, unknown> | null;
	summary: string;
	current_commit_sha: string;
	old_commit_sha: string | null;
}> {
	let effectiveRepoUrl = repoUrl;
	let effectiveBranch = branch;
	let oldSnapshot: AnalyzeRepositoryResult['snapshot'] | null = null;
	let oldDetectionResult: DetectionResult | null = null;

	if (submissionId) {
		// Support both documentId and submissionId for backward compatibility
		const document = await supabase
			.from('documents')
			.select('id, repo_id')
			.eq('id', submissionId)
			.single();

		if (document && document.data) {
			// Get repo details
			const { data: repo } = await supabase
				.from('workspace_repos')
				.select('repo_url, default_branch')
				.eq('id', document.data.repo_id)
				.single();

			if (repo) {
				effectiveRepoUrl = repo.repo_url || effectiveRepoUrl || null;
				effectiveBranch = repo.default_branch || effectiveBranch || 'main';

				// Get file hashes from repo_file_summaries for old snapshot
				const normalizedRepoId = `github.com/${parseRepoUrl(repo.repo_url || '')?.owner}/${parseRepoUrl(repo.repo_url || '')?.repo}`;
				const { data: summaries } = await supabase
					.from('repo_file_summaries')
					.select('file_path, file_hash')
					.ilike('repo_id', normalizedRepoId)
					.eq('branch', effectiveBranch);

				if (summaries) {
					const fileShas: Record<string, string | null> = {};
					summaries.forEach(s => {
						fileShas[s.file_path] = s.file_hash;
					});
					oldSnapshot = { fileShas } as any;
				}
			}
		}
	} else if (diagramId) {
		const diagram = await supabase
			.from('architecture_diagrams')
			.select('*')
			.eq('id', diagramId)
			.single();

		if (diagram && diagram.data) {
			effectiveRepoUrl =
				diagram.data.repo_url || effectiveRepoUrl || null;
			effectiveBranch =
				diagram.data.branch || effectiveBranch || 'main';
			oldSnapshot = diagram.data.code_snapshot || null;
			oldDetectionResult = diagram.data.detection_result || null;
		}
	}

	if (!effectiveRepoUrl) {
		throw new Error('repo_url is required for change detection');
	}

	const parsed = parseRepoUrl(effectiveRepoUrl);
	if (!parsed) {
		throw new Error(`Invalid GitHub URL: ${effectiveRepoUrl}`);
	}

	const resolvedBranch = effectiveBranch || parsed.branch || 'main';
	const octokit = await getUserOctokit(supabase, userId);

	// Get current commit SHA using cached method
	const branchData = await getCachedBranch(octokit, parsed.owner, parsed.repo, resolvedBranch);
	const currentCommitSha = branchData.commit.sha;
	const oldCommitSha = oldSnapshot?.commitSha || null;

	// Get new snapshot (analyzeRepository now uses optimized methods internally)
	const analyzeResult = await analyzeRepository({
		supabase,
		userId,
		repoUrl: effectiveRepoUrl,
		branch: resolvedBranch,
		subdir: parsed.subdir || null,
		filters: null,
	});

	const newSnapshot = analyzeResult.snapshot;

	// Use GitHub's compareCommits API to detect renames (with caching)
	let filesRenamed: Array<{ old_path: string; new_path: string }> = [];
	let filesChanged: FileChange[] = [];
	let filesAdded: string[] = [];
	let filesRemoved: string[] = [];

	if (oldCommitSha && oldCommitSha !== currentCommitSha) {
		// Use cached compareCommits which is more efficient
		try {
			const compareData = await getCachedCompareCommits(
				octokit,
				parsed.owner,
				parsed.repo,
				oldCommitSha,
				currentCommitSha
			);

			// Process each file change
			for (const file of compareData.files || []) {
				if (file.status === 'renamed') {
					// Detect rename
					const oldPath = file.previous_filename || file.filename;
					const newPath = file.filename;
					filesRenamed.push({
						old_path: oldPath,
						new_path: newPath,
					});

					filesChanged.push({
						path: newPath,
						old_path: oldPath,
						old_hash: null, // Will be populated below
						new_hash: null, // Will be populated below
						status: 'renamed',
					});
				} else if (file.status === 'added') {
					filesAdded.push(file.filename);
				} else if (file.status === 'removed') {
					filesRemoved.push(file.filename);
				} else if (file.status === 'modified') {
					filesChanged.push({
						path: file.filename,
						old_hash: null, // Will be populated below
						new_hash: null, // Will be populated below
						status: 'modified',
					});
				}
			}
		} catch (error) {
			console.warn('Failed to use compareCommits API, falling back to snapshot comparison:', error);
			// Fallback to snapshot comparison
			const codeComparison = compareCodeSnapshots(oldSnapshot, newSnapshot);

			filesChanged = codeComparison.filesChanged.map((change) => ({
				path: change.path,
				old_hash: change.oldHash,
				new_hash: change.newHash,
				status: 'modified' as const,
			}));
			filesAdded = codeComparison.filesAdded;
			filesRemoved = codeComparison.filesRemoved;
		}
	} else {
		// No old commit or same commit - use snapshot comparison
		const codeComparison = compareCodeSnapshots(oldSnapshot, newSnapshot);

		filesChanged = codeComparison.filesChanged.map((change) => ({
			path: change.path,
			old_hash: change.oldHash,
			new_hash: change.newHash,
			status: 'modified' as const,
		}));
		filesAdded = codeComparison.filesAdded;
		filesRemoved = codeComparison.filesRemoved;
	}

	// Populate file hashes for changed files
	const oldFileShas = oldSnapshot?.fileShas || {};
	const newFileShas = newSnapshot.fileShas || {};

	filesChanged = filesChanged.map(change => ({
		...change,
		old_hash: change.old_path
			? oldFileShas[change.old_path] || null
			: oldFileShas[change.path] || null,
		new_hash: newFileShas[change.path] || null,
	}));

	const codeComparison = {
		hasChanges: filesChanged.length > 0 || filesAdded.length > 0 || filesRemoved.length > 0 || filesRenamed.length > 0,
		commitChanged: oldCommitSha !== currentCommitSha,
		filesChanged,
		filesAdded,
		filesRemoved,
	};

	let architectureChanges: Record<string, unknown> | null = null;

	if (diagramId && oldDetectionResult) {
		const changedPaths = [
			...filesChanged.map((c) => c.path),
			...filesAdded,
		];

		// Use smart batch fetch for file content
		const detectionFiles = await fetchFilesContent(
			octokit,
			parsed.owner,
			parsed.repo,
			resolvedBranch,
			Array.from(new Set(changedPaths))
		);

		if (detectionFiles.length > 0) {
			const updatedDetection = detectTools(detectionFiles);
			const comparison = compareDetectionResults(oldDetectionResult, updatedDetection);

			architectureChanges = {
				tools_added: comparison.toolsAdded,
				tools_removed: comparison.toolsRemoved,
				connections_added: comparison.connectionsAdded,
				connections_removed: comparison.connectionsRemoved,
			};
		}
	}

	const summaryPieces: string[] = [];
	if (codeComparison.commitChanged) summaryPieces.push('Commit changed');
	if (codeComparison.filesChanged.length)
		summaryPieces.push(`${codeComparison.filesChanged.length} file(s) modified`);
	if (filesAdded.length)
		summaryPieces.push(`${filesAdded.length} file(s) added`);
	if (filesRemoved.length)
		summaryPieces.push(`${filesRemoved.length} file(s) removed`);
	if (filesRenamed.length)
		summaryPieces.push(`${filesRenamed.length} file(s) renamed`);

	const summary = summaryPieces.length ? summaryPieces.join('. ') : 'No changes detected';

	// Rate limit status tracked internally (removed verbose logging)

	if (submissionId) {
		// Note: Documents table doesn't have is_outdated field in the new schema
		// This information would need to be stored elsewhere or calculated on-demand
		// For now, we just update the document's updated_at timestamp
		await supabase
			.from('documents')
			.update({
				updated_at: new Date().toISOString(),
			})
			.eq('id', submissionId);
	} else if (diagramId) {
		await supabase
			.from('architecture_diagrams')
			.update({
				last_checked_at: new Date().toISOString(),
			})
			.eq('id', diagramId);
	}

	return {
		has_changes: codeComparison.hasChanges,
		commit_changed: codeComparison.commitChanged,
		files_changed: filesChanged,
		files_added: filesAdded,
		files_removed: filesRemoved,
		files_renamed: filesRenamed,
		architecture_changes: architectureChanges,
		summary,
		current_commit_sha: newSnapshot.commitSha,
		old_commit_sha: oldSnapshot?.commitSha || null,
	};
}

/**
 * Fetch file contents using optimized batch methods
 */
async function fetchFilesContent(
	octokit: Awaited<ReturnType<typeof getUserOctokit>>,
	owner: string,
	repo: string,
	branch: string,
	paths: string[]
): Promise<Array<{ path: string; content: string }>> {
	if (paths.length === 0) {
		return [];
	}

	// Use smart fetch which automatically chooses between individual calls and ZIP
	const files = await fetchFilesSmart(
		octokit,
		owner,
		repo,
		branch,
		paths,
		{ maxFileSize: 512 * 1024 }
	);

	return files.map(f => ({ path: f.path, content: f.content }));
}
